src/ScheduledUpgradeable.sol

```diff
@@ -0,0 +1,238 @@
+    /// @param upgradeCalldata The calldata to be executed during the upgrade
+    /// @param upgradeTime The timestamp after which the upgrade can be executed
+    /// @param nonce The nonce for the upgrade request
+    /// @return message The original encoded message
+    /// @return messageAsG1Bytes The byte representation of the BLS G1 point
+    /// @return messageAsG1Point The BLS G1 point representing the message
+    function contractUpgradeParamsToBytes(
+        string memory action,
+        address alreadyPendingImplementation,
+        address newImplementation,
+        bytes memory upgradeCalldata,
+        uint256 upgradeTime,
+        uint256 nonce
+    ) public view virtual returns (bytes memory, bytes memory, BLS.PointG1 memory) {
+        bytes memory message =
+            abi.encode(action, alreadyPendingImplementation, newImplementation, upgradeCalldata, upgradeTime, nonce);
+        (uint256 x, uint256 y) = contractUpgradeBlsValidator.hashToPoint(message);
+        BLS.PointG1 memory messageAsG1Point = BLS.PointG1({x: x, y: y});
+        bytes memory messageAsG1Bytes = abi.encode(messageAsG1Point.x, messageAsG1Point.y);
+        return (message, messageAsG1Bytes, messageAsG1Point);
+    }
+
+    /// @notice Converts BLS validator update parameters to a BLS G1 point and its byte representation.
+    /// @param blsValidator The address of the new BLS validator contract
+    /// @param nonce The nonce for the update request
+    /// @return message The original encoded message
+    /// @return messageAsG1Bytes The byte representation of the BLS G1 point
+    function blsValidatorUpdateParamsToBytes(address blsValidator, uint256 nonce)
+        public
+        view
+        virtual
+        returns (bytes memory, bytes memory, BLS.PointG1 memory)
+    {
+        bytes memory message = abi.encode(blsValidator, nonce);
+        (uint256 x, uint256 y) = contractUpgradeBlsValidator.hashToPoint(message);
+        BLS.PointG1 memory messageAsG1Point = BLS.PointG1({x: x, y: y});
+        bytes memory messageAsG1Bytes = abi.encode(messageAsG1Point.x, messageAsG1Point.y);
+        return (message, messageAsG1Bytes, messageAsG1Point);
+    }
+
+    // ---------------------- Internal Functions ----------------------
+
+    /// @dev Required by UUPS to restrict upgrades.
+    function _authorizeUpgrade(address) internal view virtual override {
+        require(msg.sender == address(this), ErrorsLib.UpgradeMustGoThroughExecuteUpgrade());
+    }
+
+    // ---------------------- Admin Functions ----------------------
+
+    /// @notice Updates the BLS validator contract used for validating admin threshold signatures.
+    /// @param _contractUpgradeBlsValidator Address of the new BLS validator contract
+    /// @param signature BLS signature from the current BLS validator validating the update
+    function setContractUpgradeBlsValidator(address _contractUpgradeBlsValidator, bytes calldata signature)
```

Comment:
should this require a signature from both the old committee and the new committee or just from the old one as now? (to prove ownership of the corresponding secret key, if you will)

Reply:
To add the new one, we will to pass in the public key too and for gas costs only the old one should be fine.
@CluEleSsUK @bbjubjub2494 @azixus  what do you guys think?

Reply:
imho it's not really necessary, it's the old committee's responsibility not to mess up, but if you want to be extra safe you can implement that.